platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,669 @@
+initializing itself with the correct round. Reconnect therefore is the responsibility of Execution. Consensus does not
+have to consider reconnect at all.
+
+## Modules
+
+### Gossip
+
+The Gossip module is responsible for gossiping messages between neighbors. The actual gossip implementation is not
+described here, except to say that it will be possible to define and implement both event-aware and event-agnostic
+gossip implementations either to a fully connected network or where the set of neighbors is a subset of the whole. This
+document does not dictate whether raw TCP, UDP, HTTP2, gRPC, or other network protocols are used. This will be left to
+the design documents for Gossip.
+
+![Gossip](gossip-module.png)
+
+Gossip is the only part of Consensus that communicates over the network with gossip neighbors. When Gossip is
+initialized, it is supplied a roster. This roster contains the full set of nodes participating in gossip, along with
+their metadata such as RSA signing keys, IP addresses, and so forth. The Gossip module decides which neighbors to gossip
+with (using whatever algorithm it chooses).
+
+#### Events
+
+Gossip is event-oriented, meaning that it is given events to gossip, and emits events it receives through gossip. An
+implementation of Gossip could be based on a lower-level implementation based on bytes, but at the module level, it
+works in terms of events.
+
+When the Gossip module receives events through gossip, it *may* choose to perform some deduplication before sending
+them to Event Intake, but it is not required to do so.
+
+Some gossip algorithms send events in *topological order*. A neighbor may still receive events out of order, because
+different events may arrive from different neighbors at different times. Events received by Gossip are not immediately
+retransmitted to its neighbors. **An honest node will only send valid events through gossip**. If invalid events are
+ever received, you may know the node that sent them to you is dishonest. Validating events increases latency at each
+"hop", but allows us to identify dishonest gossip neighbors and discipline them accordingly. For this reason, events
+received by Gossip are sent to Event Intake, which eventually send valid, ordered events _back_ to Gossip for
+redistribution.
+
+During execution, for all nodes that are online and able to keep up, events are received "live" and processed
+immediately and re-gossiped. However, if a node is offline and then comes back online, or is starting back up after
+reconnect, it may be missing events. In this case, the node will need to ask its neighbors for any events it is missing.
+
+For this reason, every honest node needs to buffer some events, so when its neighbors ask it for events, it is able to
+send them. The Gossip module **may** cache all non-expired events, but **must** cache all non-ancient events.
```

Comment:
The core problem is that it's not enforceable. Currently it's "optional" in the sense that nodes not following the protocol could stop doing this and get away with it. Probably will be similar after this redesign, in that honest nodes will cache ancient non-expired events, but there is no simple way to ensure that non-honest nodes do the same.

Reply:
In the new synch after birthRound, we're communicating our tolerated future birth round (future event horizon).  might as well communicate the current consensus round and the last supported non-expired round to give the full event horizon that can be synched over.    It could be a point of inference if Peer A communicated ancient non-expired events and Peer B did not for the same round.
.

Reply:
The challenge is that nodes that experience a reconnect may never acquire non-ancient expired events. So it's really tricky to know if the node that doesn't have them doesn't have them for a good reason, or if they discareded them when they should not have.

Reply:
Also, when we do reconnect with block nodes, it is possible that the node catches up by processing the block stream, and once it is up to date with that, it reinitializes Consensus and starts gossiping. In that case, being within the ancient window should be quite doable. So maybe ancient non-expired events won't be needed?

But to be honest, I think we'll always have them, just for that little extra safety / padding. We might not do 500 rounds when we revisit reconnect, maybe we'll only need 50 rounds or 100 rounds or 25 rounds extra.